javaLoginCore/src/main/java/com/hamdan/forzenbook/java/login/core/data/repository/LoginRepositoryImpl.java

```diff
@@ -0,0 +1,52 @@
+package com.hamdan.forzenbook.java.login.core.data.repository;
+
+import androidx.annotation.Nullable;
+
+import com.hamdan.forzenbook.java.login.core.data.database.LoginDao;
+import com.hamdan.forzenbook.java.login.core.data.network.LoginResponse;
+import com.hamdan.forzenbook.java.login.core.data.network.LoginService;
+
+import retrofit2.Response;
+
+public class LoginRepositoryImpl implements LoginRepository {
+    final private LoginDao dao;
+    final private LoginService service;
+
+    public LoginRepositoryImpl(LoginDao dao, LoginService service) {
+        this.dao = dao;
+        this.service = service;
+    }
+
+    @Override
+    public void requestValidation(String email) {
+        service.requestValidation(email);
+    }
+
+    @Override
+    public User getToken(@Nullable String email, @Nullable String code) throws NullTokenException, FailTokenRetrievalException {
+        if (email == null && code == null) return getTokenFromDatabase();
```

Comment:
I think a separate method for this may make more sense, but you can do that in a later ticket

Reply:
ok